tests/firedrake/regression/test_assemble_baseform.py

```diff
@@ -112,14 +112,32 @@ def test_assemble_action(M, f):
             assert abs(f.dat.data.sum() - 0.5*f.function_space().value_size) < 1.0e-12
 
 
-def test_scalar_formsum(f):
+@pytest.mark.parametrize("scale", ["float", "numpy", "ufl", "Constant", "Real"])
+def test_scalar_formsum(f, scale):
     c = Cofunction(f.function_space().dual())
     c.assign(1)
 
     q = action(c, f)
     expected = 2 * assemble(q)
 
-    formsum = 1.5 * q + 0.5 * q
+    s1 = 1.5
+    s2 = 0.5
+    if scale == "numpy":
+        s1 = np.asarray(s1)
+        s2 = np.asarray(s2)
+    elif scale == "ufl":
+        s1 = as_ufl(s1)
+        s2 = as_ufl(s2)
+    elif scale == "Constant":
+        s1 = Constant(s1)
+        s2 = Constant(s2)
+    elif scale == "Real":
+        mesh = f.function_space().mesh()
+        R = FunctionSpace(mesh, "R", 0)
+        s1 = Function(R, val=s1)
+        s2 = Function(R, val=s2)
```

Comment:
I'm not really sure an R0 weight makes sense here -- at least the result would probably also need to be R0. But maybe that's not for this PR.

Reply:
I think we need to support this case if we want to be able to differentiate/take adjoints with respect to weights in a `FormSum`.

Reply:
Makes sense. Implicitly mapping from function to field just feels a bit odd

```
one = fd.assemble(unity_r0_mesh0 * action_evaluating_to_one)
fd.assemble(unity_r0_mesh0 * dx(mesh1)  # Volume of mesh0/mesh1 intersection?
fd.assemble(one * dx(mesh1))  # Volume of mesh1?
```